### DISCUSSION TITLE
Help I have a few more questions besides the Log

### DISCUSSION THREAD

**Discussion by AppleCitron** [eclipse-milo/milo#1338] [General]
I am developing an application using Milo. I have some questions for you.

1. Logger problem
The application must connect to several OPC servers. The application creates and operates multiple OpcUaClients. However, only one logger leaves logs. When a single process has multiple OpcUaClients, is there any way that each OpcUaClient can leave its own log separately?

              public class OpcUaClient implements UaClient {
              
                  public static final String SDK_VERSION =
                      ManifestUtil.read("X-SDK-Version").orElse("dev");
              
                  static {
                      Logger logger = LoggerFactory.getLogger(OpcUaClient.class);
                      logger.info("Java version: " + System.getProperty("java.version"));
                      logger.info("Eclipse Milo OPC UA Stack version: {}", Stack.VERSION);
                      logger.info("Eclipse Milo OPC UA Client SDK version: {}", SDK_VERSION);
                  }

2. Is there any way to set the channel reconnection time?

          public static ChannelFsm newChannelFsm(UaStackClient client) {
                  ChannelFsmConfig fsmConfig = ChannelFsmConfig.newBuilder()
                      .setLazy(false) // reconnect immediately
                      .setMaxIdleSeconds(0) // keep alive handled by SessionFsm
                      .setMaxReconnectDelaySeconds(16)
                      .setPersistent(true)
                      .setChannelActions(new ClientChannelActions(client))
                      .setExecutor(client.getConfig().getExecutor())
                      .setScheduler(client.getConfig().getScheduledExecutor())
                      .setLoggerName(CHANNEL_FSM_LOGGER_NAME)
                      .build();
          
                  ChannelFsmFactory fsmFactory = new ChannelFsmFactory(fsmConfig);
          
                  return fsmFactory.newChannelFsm();
              }

3. Is there any way to know the session status during Milo operation?

4. Is there any way to know if KeepAlive is working during Milo's operation?

**Comment by kevinherron** [comment]
> is there any way that each OpcUaClient can leave its own log separately?

No, this is kind of just how Java logging works in general. 


> Is there any way to set the channel reconnection time?

Reconnection happens immediately and then exponentially backs off until it's ~16 seconds between attempts. It's not configurable right now - what would you change about it?

> Is there any way to know the session status during Milo operation?

Add a `SessionActivityListener` that receives `onSessionActive` and `onSessionInactive` callbacks.

> Is there any way to know if KeepAlive is working during Milo's operation?

No, but I can look at adding some kind of success/failure callback for this in the `dev/1.0` branch. It's never been something I've had a need for in the past, but I supposed it could be useful.

**Comment by AppleCitron** [comment]
Thank you for your response.

1. Are there any plans to change the log creation so that it can be generated using an ID value or without using a static variable?

2. Are there any plans to allow users to customize the maximum reconnect time?
